### PR TITLE
Priority queue

### PR DESCRIPTION
**Pull request by paulidale**
Implementation of a priority queuing mechanism and some time finctions.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
One other possibility is to include a `OSSL_TIMER_QUEUE` pointer in the `OSSL_TIMER_EVENT` structure.  This would allow the `OSSL_TIMER_QUEUE` argument to be dropped from most of the calls.  I'd keep it for the `ossl_tq_schedule` call (which would then overwrite the field in the structure).  I did look at this and it felt _wrong_ for some reason.

**Comment by levitte** [comment]
> One other possibility is to include a `OSSL_TIMER_QUEUE` pointer in the `OSSL_TIMER_EVENT` structure. This would allow the `OSSL_TIMER_QUEUE` argument to be dropped from most of the calls. I'd keep it for the `ossl_tq_schedule` call (which would then overwrite the field in the structure). I did look at this and it felt _wrong_ for some reason.

Yeah, I don't think that's the right way to go...

File: include/internal/timer_queue.h
Comment (slontis) [inline include/internal/timer_queue.h:59]: Be consistent with your blocks for one liners..

File: ssl/timer_queue.c
Comment (slontis) [inline ssl/timer_queue.c:109]: This being NULL seems like an error currently?

File: ssl/timer_queue.c
Comment (slontis) [inline ssl/timer_queue.c:106]: Why is this a while loop?

File: ssl/priority_queue.c
Comment (slontis) [inline ssl/priority_queue.c:94]: 1.5 is also less :)  ((A+2A) >> 1)

File: ssl/priority_queue.c
Comment (slontis) [inline ssl/priority_queue.c:91]: Maybe more this if above the loop and just return after line 98?

File: ssl/timer_queue.c
Comment (paulidale) [reply ssl/timer_queue.c:106]: It is meant to handle all events that are due.

It would be easy enough to only process one event per call.

I'm not sure which is better.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:94]: That risks overflow and isn't as close to optimal as 8/5.
Understand your numerics.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:91]: No, this is detecting a hard overflow on the size.

Multiple expansions are not possible currently, but if they ever are, this needs to be here.

q.v. the stack code.

File: ssl/priority_queue.c
Comment (mattcaswell) [inline ssl/priority_queue.c:264]: This is libssl so using CRYPTO_R_* seems somehow inappropriate.

File: ssl/priority_queue.c
Comment (mattcaswell) [inline ssl/priority_queue.c:268]: Isn't this really a caller error...so I'm wondering if internal error is the right reason code here. Since this is an internal API, I guess maybe its ok??

File: ssl/priority_queue.c
Comment (mattcaswell) [inline ssl/priority_queue.c:277]: Can this underflow?

File: ssl/priority_queue.c
Comment (mattcaswell) [inline ssl/priority_queue.c:306]: CRYPTO_R_ in libssl?

File: test/priority_queue_test.c
Comment (mattcaswell) [inline test/priority_queue_test.c:22]: This seems like a problem for when we start using this. Should the unused functions be marked with `ossl_unused`?

File: test/priority_queue_test.c
Comment (mattcaswell) [reply test/priority_queue_test.c:25]: If we really need this: add space after # and before pragma

File: test/priority_queue_test.c
Comment (paulidale) [reply test/priority_queue_test.c:69]: It's the number of items being removed rather than a Boolean flag.

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:62]: Yes, I'm willing to live with this :)  Or not live with it as the case may be.
This is internal at the moment so we could switch to a structure or int128 or whatever if we decide to make it public.

File: doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod
Comment (paulidale) [reply doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod:105]: Fixzed.

File: include/internal/time.h
Comment (paulidale) [reply include/internal/time.h:43]: I did it this way to begin with and it was pretty horrible to read:

`(t % OSSL_TIME_SECOND) / (OSSL_TIME_SECOND / 1000000)`

I'm not too concerned about needing to change things if we change the definition of `OSSL_TIME_SECOND`.  But if you think it better, I can.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:277]: Only if there are no entries to remove.  I.e. it's a programming error.  In debug builds, the asserts will catch this.  For non-debug builds, all kinds of badness would occur.  I'll add a check to be sure.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:268]: Not sure what's best.  Not raising any errors would be the alternative: the caller should check and catch the NULL return & raise something more interesting.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:264]: I'll switch to ERR_R_INTERNAL_ERROR for the moment.

File: ssl/time.c
Comment (paulidale) [reply ssl/time.c:31]: `return ((int64_t)now.ul) * (OSSL_TIME_SECOND / 10000000);` is less clear IMO.  There are a lot of zeros to count :grinning:

File: test/priority_queue_test.c
Comment (paulidale) [reply test/priority_queue_test.c]: Hmmm, it works without this.  So removed.

File: ssl/priority_queue.c
Comment (t8m) [reply ssl/priority_queue.c:261]: `ERR_R_PASSED_INVALID_ARGUMENT` seems better

File: ssl/priority_queue.c
Comment (t8m) [inline ssl/priority_queue.c:235]: Same as below

File: ssl/priority_queue.c
Comment (t8m) [inline ssl/priority_queue.c:299]: Again ERR_R_PASSED_INVALID_ARGUMENT or ERR_R_PASSED_NULL_PARAMETER

File: ssl/priority_queue.c
Comment (t8m) [inline ssl/priority_queue.c:336]: Same here

File: ssl/time.c
Comment (t8m) [inline ssl/time.c:38]: Isn't this missing some casts to uint64_t otherwise this will overflow at least on some platforms?

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/time.c
Comment (paulidale) [reply ssl/time.c:38]: Yes :(
Will add

File: ssl/priority_queue.c
Comment (t8m) [inline ssl/priority_queue.c:277]: If we are doing ERR_R_INTERNAL_ERROR return for empty pq, should we also unconditionally check here and return with the same error if the element is not used?

Comment (paulidale) [reply ssl/priority_queue.c:277]: These checks are only meaningful in debug mode.  They introduce an overhead.

It would be quite reasonable to not raise errors for the empty check above or even not have the checks at all.  All of this is completely internal and we don't mandate parameter checking for internal APIs.  I tend to code defensively and include them but they are easy to remove.

Comment (t8m) [reply ssl/priority_queue.c:277]: If I understand the ASSERT_ checks correctly, they are just internal consistence checks mostly which means they are OK to keep as asserts. But this particular place is different - if an external caller calls ossl_pqueue_remove() with elem that is not in the pqueue, it will result in assert in debug build and UB in non-debug build. IMO that is wrong and the public (rest of the library) facing API calls should tolerate calls where pop/remove is called on empty pq or a pq with missing item to be removed.

I'd probably even drop the ERR_R_INTERNAL_ERROR call in these cases.

Comment (paulidale) [reply ssl/priority_queue.c:277]: Currently, the information isn't maintained in non-debug tools to make this check.  It could be included but it seems like a fair bit of overhead for one check.

I suspect hitting an unused item is less likely than passing the wrong queue in :grin:

There is one internal error I think should be kept (the return check from compute_pqueue_growth), this won't be fired by a bug in the caller.

I'm almost tempted to remove all the argument and sanity run time checks rather than just the errors.

**Comment by paulidale** [comment]
Ping for reviews.

**Comment by paulidale** [comment]
Still awaiting review

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:30]: I think we should clarify that OSSL_TIME can represent a duration of time, or a point in time (representated as a duration relative to some contextually understood epoch).

We should probably also guarantee a minimum precision. I'm open about whether that precision is 1us or 1ns.

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:56]: I would leave out the word epoch and just write 0 here. Less confusing.

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:61]: I think there's some confusion here about whether OSSL_TIME represents a duration or a point in time.

For example, you have a function to convert to a timeval. timeval can represent a duration or a point in time. `gettimeofday` outputs a point in time timeval, but e.g. `select` takes a timeval representing a duration, the timeout.
 Like `struct timeval`, `struct timespec` can also represent a duration (`ppoll` timeout) or a point in time (`clock_gettime`). In the latter case though there are multiple possible clocks (epochs) so the `struct timespec` only has semantic meaning in the context of some contextually known epoch, rather than a specific fixed epoch.

I believe OSSL_TIME should be usable for representing both points in time and durations as this is the usual convention with POSIX.

I would probably put something like "OSSL_TIME can represent a duration, or a point in time. Where it is used to represent a point in time, it does so by expressing a duration relative to some reference epoch. The OSSL_TIME structure itself does not contain information about the epoch used; thus, interpretation of an OSSL_TIME requires that the epoch it is to be interpreted relative to is contextually understood."

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:19]: There's a valid question here as to what clock we want this to be. See clock_gettime(2). Since we are using this to schedule QUIC events we probably want something like `CLOCK_MONOTONIC` or `CLOCK_MONOTONIC_RAW` rather than `CLOCK_REALTIME`. Because of this we should document that this is not suitable for wall-clock time uses (e.g. checking certificate expiry).

The epoch used by this function probably doesn't need to be defined. It isn't for CLOCK_MONOTONIC.

The documented contract for this API can be deliberately vague. Something like: "This function returns the current time relative to an epoch which is undefined but unchanging for at least the duration of program execution. The time returned is monotonic and need not represent wall-clock time. The time returned by this function is useful for scheduling timeouts, deadlines and recurring events, but due to its undefined epoch and monotonic nature, is not suitable for other uses."

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:62]: I don't think we should fix this as part of the defined API, or if we do we should choose a concrete representation of OSSL_TIME rather than being cagey about its representation, as this basically implies its representation. Rather we should specify a lower bound on the maximum representable value. I'd also express it as a duration first without mentioning any particular epoch ("This value can represent a duration of at least ~584 years.").

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:75]: propose s/the Epoch/0/

File: doc/internal/man3/OSSL_TIME.pod
Comment (hlandau) [inline doc/internal/man3/OSSL_TIME.pod:79]: nit: s/to/in/ (maybe, subjective)

File: ssl/time.c
Comment (hlandau) [inline ssl/time.c:35]: This is `CLOCK_REALTIME` so it can jump around, therefore we should not use it to schedule QUIC timer events. `ifdef CLOCK_MONOTONIC` is probably a decent way to detect if the platform supports `clock_gettime`.  `gettimeofday` might be a suboptimal but tolerable fallback if it's not available.

The Windows equivalent you want is `QueryPerformanceCounter` (which is what C++'s `std::chrono::steady_clock` uses). For MacOS, `clock_gettime(CLOCK_UPTIME_RAW)` apparently. Source: https://github.com/ros2/rcutils/issues/43

File: include/internal/priority_queue.h
Comment (hlandau) [inline include/internal/priority_queue.h:50]: Does not match documentation?

File: include/internal/priority_queue.h
Comment (hlandau) [inline include/internal/priority_queue.h:17]: No definition for PRIORITY_QUEUE_ELEM_OF

File: ssl/priority_queue.c
Comment (hlandau) [reply ssl/priority_queue.c:261]: I would probably just document it as a precondition in the manpage that the element must be in the queue, and that undefined behaviour results if it isn't.

File: test/priority_queue_test.c
Comment (hlandau) [inline test/priority_queue_test.c:58]: nit: space after opening brace, so should be one before closing brace (or remove it)

**Review by hlandau — CHANGES_REQUESTED** [review]

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:20]: I'm not sure about `struct timespec`, it would need conditioning.  Another one that might be useful is `time_t`.  I didn't think about the reverse versions which are a good idea.  The big caveat is that these are internal right now and we don't need the other forms.  The only things libssl wants is the current time as a `struct timeval`, so that's what I did.

If we ever made them public, we'd want to revisit the conversions made available.

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:39]: It's a little more nuanced, I'm relying on being able to sort times.  An overflow must therefore go to the largest possible value and an underflow to zero.  Hence the saturating arithmetic.  Is there a need to document this behaviour?

TIME_NEVER isn't a great name :sob:

File: test/priority_queue_test.c
Comment (paulidale) [reply test/priority_queue_test.c:58]: done.

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:79]: We seem to alternate between _This functionality was added to_ and _These functions were added in_.
Yeah for consistency...

File: doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod
Comment (paulidale) [reply doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod:83]: Missed documentation updates elsewhere :(
This is correct :)

File: doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod
Comment (paulidale) [reply doc/internal/man3/DEFINE_PRIORITY_QUEUE_OF.pod:86]: The only use it could be put to is calling remove.  pop already does that.

File: ssl/priority_queue.c
Comment (paulidale) [reply ssl/priority_queue.c:261]: Good idea, I've add: _It is undefined what happens if an attempt to remove an element that isn't in the queue is made._

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:30]: `OSSL_TIME_SECOND` defines the precision, I'm not sure we need to be more specific than that.  Maybe if it were a public API...

I've added _It can represent both a specific point time or a duration._ to attempt to clarify the definition of `OSSL_TIME`.

Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:30]: `OSSL_TIME_SECOND` defines the precision.  I think leaving the minimum unspecified is sufficient for now.  If this API goes public, we might want to reconsider.

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:19]: +1 on the text which I've included.

Which clock we use will be an interesting discussion.  Currently, I've used what libssl uses for DTLS and I agree it's not ideal.  It's not a straightforward substitution, `gettimeofday(2)` is more widely available than `clock_gettime(2)` and doesn't require `-lrt` on the link line.  My guess is that we will want to support both falling back to the less accurate as required.

Various operating systems also expose higher resolution timers (Solaris, HP/UX and AIX from memory) which is an additional complexity I'd not to not dive into :grinning:

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:62]: Good point, changed.

File: ssl/time.c
Comment (paulidale) [reply ssl/time.c:35]: `CLOCK_MONOTONIC_RAW` would be ideal I think.  The approach I've used in the past is a series of preprocessor checks for the various clocks in preferential order:
```c
#if defined(CLOCK_MONOTONIC_RAW)
#elif defined(CLOCK_MONOTONIC)
#elif defined(CLOCK_REALTIME)
#else
   // gettimeofday
#endif
```
Usually with some OS specific interventions because everyone does this slightly differently.

I think the timer used should be a separate discussion probably involving a long chunk of OTC time.

**Comment by paulidale** [comment]
Good feedback @hlandau

File: doc/internal/man3/OSSL_TIME.pod
Comment (kaduk) [reply doc/internal/man3/OSSL_TIME.pod:44]: `TIME_INFINITY` (not even a lie, with saturating arithmetic)

Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:44]: Going with the infinite.

File: ssl/time.c
Comment (paulidale) [reply ssl/time.c:35]: Refer to #18519 for the decision.  I'll mark this as resolved since I think the specific choice is separate.

File: doc/internal/man3/OSSL_TIME.pod
Comment (mattcaswell) [inline doc/internal/man3/OSSL_TIME.pod:14]: This line doesn't seem to add anything. I'd remove it.

File: include/internal/time.h
Comment (mattcaswell) [inline include/internal/time.h:18]: Typo: precision??

File: ssl/priority_queue.c
Comment (mattcaswell) [inline ssl/priority_queue.c:18]: Should start with blank line:

````
/*
 * Fundamental operations:
````

**Review by mattcaswell — APPROVED** [review]
Approved subject to the minor issues below being resolved.

File: doc/internal/man3/OSSL_TIME.pod
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod:14]: Isn't this type the entire point of this documentation?
Without this line, what is an `OSSL_TIME`?

I didn't want to tie us down to a specific type, it might need to be a structure rather than an integer at some point...

Comment (mattcaswell) [reply doc/internal/man3/OSSL_TIME.pod:14]: > Without this line, what is an OSSL_TIME?

We don't do this for most other types, so why are we making an exception for this one?  Most of our types are opaque structures, but we don't specify that in the documentation.

Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod]: It's internal documentation, it's not critical until it's made public :smile:

Comment (mattcaswell) [reply doc/internal/man3/OSSL_TIME.pod]: Ok - fair enough.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the feedback.

**Pull request closed without merging**
